rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md

```diff
@@ -1 +1,157 @@
-In review, see <https://github.com/nervosnetwork/rfcs/pull/238>, <https://github.com/nervosnetwork/rfcs/pull/236>, and <https://github.com/nervosnetwork/rfcs/pull/237>
+---
+Number: "0032"
+Category: Consensus (Hard Fork)
+Status: Draft
+Author: Ian Yang
+Organization: Nervos Foundation
+Created: 2021-04-26
+---
+
+# CKB VM Version Selection
+
+## Abstract
+
+This RFC proposes a mechanism to choose the CKB VM version after the hard fork. What's included in the new CKB VM version is beyond the scope of this document.
+
+## Motivation
+
+It's essential to keep improving CKB VM because it is the computation bottleneck of the whole network. The upgrade packages can improve the performance, bring bug fixings and add new RISC-V extensions. However the upgrade should not break the old code, users must have the opt-in option to specify the VM version.
+
+This RFC proposes a general mechanism that determines how the CKB node chooses the CKB VM version for a transaction script group.
+
+## Specification
+
+The CKB VM upgrades via hard fork. The next scheduled hard fork is ckb2021, which is activated since a specific epoch. For all the transactions in the blocks before the activation epoch, they must run the CKB VM version 0 to verify all the script groups. In these transactions, the `hash_type` in cell lock and type script must be either 0 or 1 in the serialized molecule data.
+
+After ckb2021 is activated, CKB nodes must choose the CKB VM version for each script group. The allowed values for the `hash_type` field in the lock and type script are 0, 1, and 2. Cells are sorted into different groups if they have different `hash_type`. According to the value of `hash_type`:
+
+* When the `hash_type` is 0, the script group matches code via data hash and will run the code using the CKB VM version 0.
+* When the `hash_type` is 1, the script group matches code via type script hash and will run the code using the CKB VM version 1.
+* When the `hash_type` is 2, the script group matches code via data hash and will run the code using the CKB VM version 1.
+
+| `hash_type` | matches by       | VM version |
```

Comment:
I would recommend a short section breaking down the bits of `hash_type`. People will wonder why `0` in hash type means `data hash + VM version 0`, but `2` in hash type means `data hash + VM version 1`(why not use 1 here).

Reply:
I'll explain why we choose such encoding pattern.

However, I will not mention the bits level encoding of `hash_type` because:

1. We have not decided yet how to encode `data hash + VM version 2` and following versions
2. We want developers use this field as an enum, instead of a continuous version number plus a type flag.